### ISSUE TITLE
on disk caching weird behavior

### ISSUE THREAD

**Issue by ysholqamy** [Alamofire/AlamofireImage#22]
Hello, I just encountered an issue will using the pod. First of all, some images are fetched then written to disk at the correct path which is {app_identifier}/com.alamofire.imagedownloader/NsCachedData which is perfect as these images are not even downloaded again when the app is in offline mode. But the issue is that sometimes images are written to {app_identifier}/NsCachedData which I believe is the path for NSURLCache.sharedURLCache and not the path specified in the cache used by alamofireImage so these images are not seen by ImageDownloader and are re downloaded if they are not found in-memory. Also sometimes images are not even written to the disk which is weird.

All these Images where fetched using the same method with the same approach tried removing their filters and nothing happened still the same behavior. I used ImageDownloader.defaultInstance.downloadImage method to fetch these Images.

Any help will be appreciated, thanks In advance!

**Comment by cnoon** [comment]
How are you using the `ImageDownloader`? Could you post all relevant sample code? Are you only using the `UIImageView` extension?

**Comment by ysholqamy** [comment]
No I am not using the UIImageView extension. I am using ImageDownloader.defaultInstance.downloadImage(url, filter, completion) function

The code is a straight forward usage of downloadImage function

**Comment by ysholqamy** [comment]
I also noticed that the image which wont be written to disk had an expiry date header in its response, that's the only difference between those images beside their extensions.

**Comment by cnoon** [comment]
Could you post your initialization logic for the `ImageDownloader`? I need to know exactly how you are setting everything up because this is going to come down to how you have the `NSURLCache` configured.

**Comment by kcharwood** [comment]
Would you be able to provide any example URKs as well? I'd like to see the response cache headers.

**Comment by kcharwood** [comment]
URLs*

**Comment by cnoon** [comment]
+1 ☝🏽

**Comment by ysholqamy** [comment]
Sorry for the late reply. I did not initialize ImageDownloader myself, I just used the defaultInstance. I dont have the laptop on me right now so I will post the URLS later today.

**Comment by ysholqamy** [comment]
here is url http://konnecti.s3.amazonaws.com/system/defaults/avatar.png

**Comment by cnoon** [comment]
Which OS @ysholqamy?

**Comment by ysholqamy** [comment]
Tested with iOS 9.0 and 8.4

**Comment by cnoon** [comment]
Okay, I've been debugging this and have a bunch of information. I can't reproduce any of the behavior @ysholqamy has reported, but here goes.

### Headers

```bash
cnoon:~$ curl -s -D - http://konnecti.s3.amazonaws.com/system/defaults/avatar.png -o /dev/null
HTTP/1.1 200 OK
x-amz-id-2: B1WPwkdmpZLuqykEbVhrUCdk5XNiA5EJDD4uWPARJGB4i1Q1Amc3fPtysjY4lIKEx+Dske3RZS0=
x-amz-request-id: 66068F1546DA94BB
Date: Tue, 29 Sep 2015 15:19:55 GMT
Last-Modified: Wed, 26 Aug 2015 10:50:32 GMT
ETag: "c208c8514e6d76ea3afc3ad85726bff5"
Accept-Ranges: bytes
Content-Type: image/png
Content-Length: 9173
Server: AmazonS3
```

Replaced the `ImageViewController` URL in the iOS Example of AlamofireImage with the URL that @ysholqamy attached `http://konnecti.s3.amazonaws.com/system/defaults/avatar.png`. I also had to disable ATS to get it working. Here are some screenshots of the database tables.

![screen shot 2015-09-29 at 8 17 19 am](https://cloud.githubusercontent.com/assets/169110/10168800/4b735b6a-6683-11e5-917f-6f15e14895a8.png)


![screen shot 2015-09-29 at 8 17 22 am](https://cloud.githubusercontent.com/assets/169110/10168806/54b8f234-6683-11e5-9bbb-e82419c18850.png)


![screen shot 2015-09-29 at 8 17 25 am](https://cloud.githubusercontent.com/assets/169110/10168814/58f1eaae-6683-11e5-953e-5ce25c977180.png)


Everything is being cached as expected with the exception of the avatar image. For some reason, it's not being stored on-disk.


![screen shot 2015-09-29 at 8 24 52 am](https://cloud.githubusercontent.com/assets/169110/10168902/97c34322-6683-11e5-84cd-0814b952a13f.png)

I don't see anything in the headers that would stop `NSURLCache` from storing the data locally. That seems a bit odd to me, but otherwise all the data is being stored exactly where it should be. How about you @kcharwood? Anything stand out to you?

@ysholqamy are you sure you're not calling that URL anywhere for something outside AlamofireImage? It seems pretty impossible to me that the ImageDownloader is putting that file anywhere other than I just posted.

**Comment by ysholqamy** [comment]
The response for the URL I provided was never cached on disk, other images were written but directly in a folder called NsCachedData which is a sibling of alamofireImage folder, other images were written the right place which you just posted

**Comment by ysholqamy** [comment]
What is odd too, is the headers you just posted. When I make a request for this URL a header named expiry-date is included in the headers which are missing the headers you just posted which is weird

**Comment by ysholqamy** [comment]
$ curl -s -D - http://konnecti.s3.amazonaws.com/system/defaults/avatar.png
HTTP/1.1 200 OK
x-amz-id-2: lwvH6LpcEsdsqRqriUUqVqPPTTQZeqNBD/NO7NrA09d+yMn0Gu5GPexsYLQGje3h
x-amz-request-id: EB7DAC04446F0236
Date: Tue, 29 Sep 2015 15:39:32 GMT
Last-Modified: Wed, 26 Aug 2015 10:50:32 GMT
ETag: "c208c8514e6d76ea3afc3ad85726bff5"
Accept-Ranges: bytes
Content-Type: image/png
Content-Length: 9173
Server: AmazonS3
Expires: Wed, 30 Sep 2015 15:30:59 GMT
Connection: keep-alive

**Comment by kcharwood** [comment]
I think the response has to have a `Cache-Control` header in order for NSURLCache to pick it up. Do those other cached requests have that header?

**Comment by kcharwood** [comment]
Check NSHipster

Cache-Control is required header. You can override the cache response block on the session manager to force the asset to be cached if you would like.

**Comment by cnoon** [comment]
Headers:

```swift
<NSHTTPURLResponse: 0x7f86e2f5c400> { URL: https://secure.gravatar.com/avatar/14d7e86e7f713d362f70a89964b840ff?d=identicon&r=pg&f=y&s=382 } { status code: 200, headers {
    "Accept-Ranges" = bytes;
    "Access-Control-Allow-Origin" = "*";
    "Cache-Control" = "max-age=300";
    "Content-Length" = 105220;
    "Content-Type" = "image/png";
    Date = "Tue, 29 Sep 2015 15:43:45 GMT";
    Expires = "Tue, 29 Sep 2015 15:48:45 GMT";
    "Last-Modified" = "Wed, 11 Jan 1984 08:00:00 GMT";
    Link = "<https://www.gravatar.com/avatar/14d7e86e7f713d362f70a89964b840ff?d=identicon&r=pg&f=y&s=382>; rel=\"canonical\"";
    Server = nginx;
    "Source-Age" = 0;
    Via = "1.1 varnish";
    "x-varnish" = 4284027326;
}
```

**Comment by cnoon** [comment]
Ah that would explain my findings @kcharwood. I didn't realize it was required for NSURLCache to store it on-disk.

**Comment by cnoon** [comment]
Well @ysholqamy I don't really know what's going on here. It appears from our end that everything is working exactly as expected. Unless you can provide us with some additional info, I'm afraid we're going to have to close this out.

**Comment by ysholqamy** [comment]
HTTP/1.1 200 OK
x-amz-id-2: aHCuK3QoaDDYrALnsz7W2Iog0uUybI/u+Zdz42anS/2ITZk1D59zuXGWSip55HSm
x-amz-request-id: 6C9BE2B57D3F4A54
Date: Tue, 29 Sep 2015 15:45:22 GMT
Last-Modified: Sat, 26 Sep 2015 22:27:19 GMT
ETag: "1171a3125089a42deeaf05021a051b7f"
Accept-Ranges: bytes
Content-Type: image/jpeg
Content-Length: 59099
Server: AmazonS3

**Comment by ysholqamy** [comment]
the headers I just posted is the response headers for one of the images which are cached on-disk and they dont include cache-control header

**Comment by cnoon** [comment]
You should print the `NSHTTPURLResponse` out to verify. It appears we're seeing discrepancies between cURL headers and the `NSHTTPURLResponse` headers.

**Comment by kcharwood** [comment]
You can also use `setDataTaskWillCacheResponseBlock:` on the session manager to force the cache.

**Comment by ysholqamy** [comment]
These are the headers from NSHTTPURLResponse for one of the images which are are cached correctly
  
     { status code: 200, headers {
    "Accept-Ranges" = bytes;
    "Content-Length" = 59099;
    "Content-Type" = "image/jpeg";
    Date = "Tue, 29 Sep 2015 15:56:16 GMT";
    Etag = "\"1171a3125089a42deeaf05021a051b7f\"";
    "Last-Modified" = "Sat, 26 Sep 2015 22:27:19 GMT";
    Server = AmazonS3;
    "x-amz-id-2" = "Va27Ee+C8sdhOoxDmbu+5WrPq6Gd8cawPpgbjYWoe6lLpsdTSCD3MvLtX9Orjqx9pChjJFWnvX4=";
    "x-amz-request-id" = 2DE8630C1AE64608; }} 

and these are the headers from NSHTTPURLResponse for the image which wont be cached on-disk

      { status code: 200, headers {
    "Accept-Ranges" = bytes;
    Connection = "keep-alive";
    "Content-Length" = 9173;
    "Content-Type" = "image/png";
    Date = "Tue, 29 Sep 2015 16:01:38 GMT";
    Etag = "\"c208c8514e6d76ea3afc3ad85726bff5\"";
    Expires = "Wed, 30 Sep 2015 15:53:05 GMT";
    "Last-Modified" = "Wed, 26 Aug 2015 10:50:32 GMT";
    Server = AmazonS3;
    "x-amz-id-2" = "nEr7rx/xHdtFolp5V2LWQgaON+DvtVV9z1v1LpFQ1VOMQmn3dxDr71Mmhsblalxt";
    "x-amz-request-id" = 0183F35886F5D907;} }

**Comment by kcharwood** [comment]
The `expires` tag is interesting... 

Some of these problems are coming back to me now. I think it has something along the lines of if `cache-control` isn't present, the OS makes some guesses on whether it should actually cache the image or not. Since that response has an expires tag, the OS may be deciding its not worth to cache that item for "only" 24 hours.

Undocumented, black magic.

**Comment by ysholqamy** [comment]
Yea I think the expires tag is the problem, just read an article about it which says that OS makes some guesses for how long this image will be valid but never mentioned that the OS wont cache it in the first place.

Also here is screenshot of the device content which I am testing on, it shows the image is written outside of alamofireImage folder 
![screen shot 2015-09-29 at 6 07 29 pm](https://cloud.githubusercontent.com/assets/1161468/10170449/38ddd686-66d5-11e5-804c-b37de8b0203b.png)

**Comment by ysholqamy** [comment]
also the documentation says that the response will be nil if the image was found in the local cache whether on disk or in memory, but actually I just tried fetching the image cached on disk while the device was in airplane mode, the image was fetched successfully from the disk but the response was not nil which is weird because the yesterday the image was not even fetched when it was written the sibling folder of alamofireimage folder.

**Comment by ysholqamy** [comment]
And btw I would like to help, I think the project is awesome and really useful, but I could not find any contributing guidelines. Great job :)

**Comment by cnoon** [comment]
Have you verified the `8BEE445F...` UUID file is actually an image that you downloaded using AlamofireImage? I would at this point almost bet money that it is not an image but instead of JSON payload of some kind.

**Comment by kcharwood** [comment]
The response is nil if it's returned from the in memory cache. If the response isn't nil, and you're in airplane mode, then it was returned from the NSURLCache

**Comment by ysholqamy** [comment]
Yea I downloaded the package content and that stored file was the image

**Comment by cnoon** [comment]
Okay, well I can't speak for @kcharwood, but at this point I'm out of ideas. Would it be possible for you to put together a sample app that demonstrates the behavior? Otherwise I'm not sure what else we can do to help.

**Comment by ysholqamy** [comment]
Ok I will put a sample app tonight. Thanks alot

**Comment by kcharwood** [comment]
I would assume its that `Expires` tag.

**Comment by cnoon** [comment]
I'm going to close this issue out since there is no identified issue with AlamofireImage here. @ysholqamy if you end up posting a sample project that can demonstrate an issue, I'll be happy to re-open and investigate further.

**Comment by Stakenborg** [comment]
I'm having the same issue with my cached images appearing outside of the com.alamofire.imagedownloader folder. I've tried both the `UIImageView` extension and a default initialized `ImageDownloader`, but I'm seeing the same issue with both. I even tried using the Gravatar example posted here, which also wound up in this other folder. Trying to figure out why this is happening, but so far I'm at a loss.

Here's a shot of my cache folder:
http://puu.sh/o8nEt/5eb3d9d364.png

**Comment by Stakenborg** [comment]
I was able to get this...sort of...working. Seems for some reason `NSURLCache` is only playing nice with a nil diskPath. If I specify that in the configuration when creating my `ImageDownloader` it works, but if I specify any diskPath the path gets created, but never written to.

To make things even more strange, I tried to manually store/retrieve with my own `NSURLCache` object, which worked, but again created the diskPath directory, but seems to read/write from my app identifier root directory.

I wish I could help more, but now that this is working well enough for me, I've got to move on with my project. I'll let everyone know if I discover anything else.
